processing/src/main/java/org/apache/druid/segment/join/JoinVirtualColumnSplit.java

```diff
@@ -0,0 +1,161 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.segment.join;
+
+import org.apache.druid.query.filter.Filter;
+import org.apache.druid.segment.VirtualColumn;
+import org.apache.druid.segment.VirtualColumns;
+import org.apache.druid.segment.column.RowSignature;
+import org.apache.druid.segment.join.filter.JoinFilterSplit;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+public class JoinVirtualColumnSplit
+{
+  /**
+   * Split {@link VirtualColumns} into those which can have inputs which exist on the base table and those which can
+   * be deferred until after the join due to not being present in the base table and thus all null
+   */
+  public static JoinVirtualColumnSplit split(
+      RowSignature baseTableSignature,
+      VirtualColumns virtualColumns,
+      JoinFilterSplit filterSplit
+  )
+  {
+    final Comparator<VirtualColumn> sorter = Comparator.comparing(VirtualColumn::getOutputName);
+    final SortedSet<VirtualColumn> preJoinVirtualColumns = new TreeSet<>(sorter);
+    final SortedSet<VirtualColumn> postJoinVirtualColumns = new TreeSet<>(sorter);
+    final Set<String> baseColumnNames = new HashSet<>(baseTableSignature.getColumnNames());
+    final Set<String> baseFilterColumns = filterSplit.getBaseTableFilter()
+                                                     .map(Filter::getRequiredColumns)
+                                                     .orElse(Collections.emptySet());
+    final Set<String> unresolvedNames = new HashSet<>();
+
+    final Set<VirtualColumn> unresolvedSet = new HashSet<>();
+    // do a first pass to eliminate those which depend directly on the base table or are required by base table filters
+    for (VirtualColumn virtualColumn : virtualColumns.getVirtualColumns()) {
+      if (baseColumnNames.containsAll(virtualColumn.requiredColumns())) {
+        // virtual column depends on physical columns, or virtual columns which have already been added to the 'pre' set
+        preJoinVirtualColumns.add(virtualColumn);
+        baseColumnNames.add(virtualColumn.getOutputName());
+      } else if (baseFilterColumns.contains(virtualColumn.getOutputName())) {
+        // virtual column is required by a base table filter clause
+        preJoinVirtualColumns.add(virtualColumn);
+        baseColumnNames.add(virtualColumn.getOutputName());
+        // also resolve all virtual column dependencies it has and add them to the base table set
+        final Set<VirtualColumn> dependencies = resolveDependencies(virtualColumns, virtualColumn);
+        for (VirtualColumn dep : dependencies) {
+          preJoinVirtualColumns.add(dep);
+          postJoinVirtualColumns.remove(dep);
```

Comment:
this seems redundant.

Reply:
oops yes, earlier i was populating `postJoinVirtualColumns` in the else